plugins/cilium-cni/eni.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/pkg/datapath/linux/route"
+
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/vishvananda/netlink"
+)
+
+func prepareENI(mac string, mtu int) (index int, err error) {
+	var links []netlink.Link
+
+	links, err = netlink.LinkList()
+	if err != nil {
+		err = fmt.Errorf("unable to list interfaces: %s", err)
+		return
+	}
+
+	for _, link := range links {
+		if link.Attrs().HardwareAddr.String() == mac {
+			index = link.Attrs().Index
+
+			if err = netlink.LinkSetMTU(link, mtu); err != nil {
+				err = fmt.Errorf("unable to change MTU of link %s to %d: %s", link.Attrs().Name, mtu, err)
+				return
+			}
+
+			if err = netlink.LinkSetUp(link); err != nil {
+				err = fmt.Errorf("unable to up link %s: %s", link.Attrs().Name, err)
+				return
+			}
+
+			return
+		}
+	}
+
+	err = fmt.Errorf("interface with MAC %s not found", mac)
+	return
+}
+
+func eniAdd(ipConfig *current.IPConfig, ipam *models.IPAMAddressResponse, conf models.DaemonConfigurationStatus) error {
+	for _, cidrString := range ipam.Cidrs {
+		_, _, err := net.ParseCIDR(cidrString)
+		if err != nil {
+			return fmt.Errorf("invalid CIDR '%s': %s", cidrString, err)
+		}
+	}
+
+	if ipam.MasterMac == "" {
+		return fmt.Errorf("ENI master interface MAC address is not set")
+	}
+
+	ifindex, err := prepareENI(ipam.MasterMac, int(conf.DeviceMTU))
```

Comment:
IIRC in the regular Cilium operation, we configure MTU on devices for the maximum possible receive packet size (eg, from the internet); then we configure routes inside the endpoint with MTU to support the maximum anticipated packet size that the underlay will accept (commit 97a1a116241ab).

I guess that in this case, we don't expect to operate ENI with an overlay because the whole point is that ENI provides direct connectivity to all pods/nodes; so in that case we don't need to worry about configuring the MTU on the routes, as we can rely on the MTU on the devices. Is that in line with your understanding?

Reply:
Correct. If a user configured encapsulation in combination with ENI IPAM then the ENI routes won't be used for any pod to pod traffic. I don't see much a reason to run in that mode though as direct routing mode has several advantages. Both modes will work fine though regarding MTU.